Add image query support to the backend microservices

dmsuehir opened the pull request:
## Description

Updates the following microservices:
* gateway: Updated to support multiple prompt formats depending on the model being used, updated send an image to the embedding service for first queries, and allow sending multiple images to the LVM.
* lvm-llava: Updated to allow multiple images to be passed in and added support for other prompt formats depending on the LVM model. The lvm-llava-svc will also limit the number of images that get passed to the LVM using MAX_IMAGES
* embedding: This microservice already allowed image input. Updated to pass that image through with the emedding output.
* retriever: Updated to pass the original image through with the output (along with the image that was found with the similarity search)

## Issues

https://github.com/opea-project/docs/blob/main/community/rfcs/24-10-02-GenAIExamples-001-Image_and_Audio_Support_in_MultimodalQnA.md

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds new functionality)

## Dependencies

No new dependencies

## Tests

Tests have been updated

dmsuehir committed:
Backend enhancements for image query capabilities for MultimodalQnA

dmsuehir committed:
Fix model name var

Signed-off-by: dmsuehir <dina.s.jones@intel.com>

dmsuehir committed:
Merge branch 'mmqna-phase2' of github.com:mhbuehler/GenAIComps into dina/image_query

dmsuehir committed:
Remove space at end of prompt

Signed-off-by: dmsuehir <dina.s.jones@intel.com>

dmsuehir committed:
Merge branch 'mmqna-phase2' of github.com:mhbuehler/GenAIComps into dina/image_query

dmsuehir committed:
Add env var for the max number of images sent to the LVM

Signed-off-by: dmsuehir <dina.s.jones@intel.com>

dmsuehir committed:
README update for the MAX_IMAGES env var

Signed-off-by: dmsuehir <dina.s.jones@intel.com>

dmsuehir committed:
Merge branch 'dina/image_query' of github.com:mhbuehler/GenAIComps into dina/image_query

dmsuehir committed:
Remove prints

Signed-off-by: dmsuehir <dina.s.jones@intel.com>

okhleif-10 committed:
Audio query functionality to multimodal backend (#8)

Signed-off-by: okhleif-IL <omar.khleif@intel.com>

* added in audio dict creation

Signed-off-by: okhleif-IL <omar.khleif@intel.com>

* separated audio from prompt

Signed-off-by: okhleif-IL <omar.khleif@intel.com>

* added ASR endpoint

Signed-off-by: okhleif-IL <omar.khleif@intel.com>

* removed ASR endpoints from mm embedding

Signed-off-by: okhleif-IL <omar.khleif@intel.com>

* edited return logic, fixed function call

Signed-off-by: okhleif-IL <omar.khleif@intel.com>

* added megaservice to elif

Signed-off-by: okhleif-IL <omar.khleif@intel.com>

* reworked helper func

Signed-off-by: okhleif-IL <omar.khleif@intel.com>

* Append audio to prompt

Signed-off-by: okhleif-IL <omar.khleif@intel.com>

* Reworked handle messages, added metadata

Signed-off-by: okhleif-IL <omar.khleif@intel.com>

* Moved dictionary logic to right place

Signed-off-by: okhleif-IL <omar.khleif@intel.com>

* changed logic to rely on message len

Signed-off-by: okhleif-IL <omar.khleif@intel.com>

* list --> empty str

Signed-off-by: okhleif-IL <omar.khleif@intel.com>
---------

Signed-off-by: Melanie Buehler <melanie.h.buehler@intel.com>
Signed-off-by: okhleif-IL <omar.khleif@intel.com>
Signed-off-by: dmsuehir <dina.s.jones@intel.com>

dmsuehir committed:
Merge branch 'mmqna-phase2' of github.com:mhbuehler/GenAIComps into dina/image_query

Signed-off-by: dmsuehir <dina.s.jones@intel.com>

mhbuehler committed:
Merge branch 'main' into mmqna-audio-query

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

okhleif-10 committed:
fixed role bug where i never was > 0

Signed-off-by: okhleif-IL <omar.khleif@intel.com>

dmsuehir committed:
Fix after merge

Signed-off-by: dmsuehir <dina.s.jones@intel.com>

okhleif-10 committed:
removed whitespace

Signed-off-by: okhleif-IL <omar.khleif@intel.com>

mhbuehler committed:
Merge pull request #13 from mhbuehler/omar/role-debug

Fixed role bug where enumeration was wrong

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

dmsuehir committed:
Fix call to get role labels

Signed-off-by: dmsuehir <dina.s.jones@intel.com>

dmsuehir committed:
Merge branch 'mmqna-audio-query' of github.com:mhbuehler/GenAIComps into dina/image_query

Signed-off-by: dmsuehir <dina.s.jones@intel.com>

dmsuehir committed:
Gateway test updates images within the conversation

Signed-off-by: dmsuehir <dina.s.jones@intel.com>

dmsuehir commented on tests/cores/mega/test_multimodalqna_gateway.py:
```diff
@@ -65,7 +66,23 @@ async def lvm_add(request: Union[LVMDoc, LVMSearchedMultimodalDoc]) -> TextDoc:
     else:
         print("request is from user.")
         text = req_dict["prompt"]
-        text = f"<image>\nUSER: {text}\nASSISTANT:"
```
It used to be that the LVM microservice would always add `<image>\n` and `USER:` to the beginning of the prompt that was created by the gateway, however I had to change that since images can now to scattered throughout the conversation (we can no longer just attach the single image to the first message in the conversation). I updated this mock of the LVM microservice to do something similar to what the actual LVM service does where it checks how many image tags already exist in the prompt, and adds extras if they are needed. The reason why we might need extras are when the retriever also gets an image from the vector store.

dmsuehir commented on tests/cores/mega/test_multimodalqna_gateway.py:
```diff
@@ -111,7 +128,7 @@ async def test_follow_up_query_service_builder_schedule(self):
         )
         # print(result_dict)
-        self.assertEqual(result_dict[self.lvm.name]["text"], "<image>\nUSER: chao, \nASSISTANT:")
+        self.assertEqual(result_dict[self.lvm.name]["text"], "USER: <image>\nchao, \nASSISTANT:")
```
Previously, the prompts had `<image>\nUSER:` but the prompt format [documented by huggingface](https://huggingface.co/docs/transformers/en/model_doc/llava) is the other way around `USER: <image>\n` and also since we will now have images interleaved in the conversation, I changed it to match the HF docs.

dmsuehir commented on tests/cores/mega/test_multimodalqna_gateway.py:
```diff
@@ -205,9 +222,75 @@ async def test_handle_request(self):
+        formats than the default LLaVA 1.5 model.
+        """
+
+        # Models to test and their expected prompts
```
Normally I would've parameterized this test with the pytest decorator, however it seems like that won't work in this case because of them subclassing unittest.IsolatedAsyncioTestCase. Apparently there's another library called parameterized that could do it, but the unittest container that this is run in doesn't have that dependency and it's easier to avoid adding a third party dependency ([stackoverflow reference for the issue](https://stackoverflow.com/questions/35561843/py-test-parametrize-test-cases-from-classes)).

So, instead I just have lists of the models and their expected prompt format, and I'm looping those to do the checks.

mhbuehler committed:
Adds unit test coverage for audio query

Signed-off-by: Melanie Buehler <melanie.h.buehler@intel.com>

dmsuehir committed:
Update test to check the returned b64 types

Signed-off-by: dmsuehir <dina.s.jones@intel.com>

dmsuehir committed:
Update test since we don't expect images from the assistant

Signed-off-by: dmsuehir <dina.s.jones@intel.com>

mhbuehler committed:
Port number fix

Signed-off-by: Melanie Buehler <melanie.h.buehler@intel.com>

mhbuehler committed:
Formatting

Signed-off-by: Melanie Buehler <melanie.h.buehler@intel.com>

mhbuehler committed:
Merge pull request #14 from mhbuehler/melanie/add_test_coverage

Adds unit test coverage for audio query

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ashahba committed:
Merge branch 'main' into mmqna-audio-query

mhbuehler committed:
Fixed place where port number is set

Signed-off-by: Melanie Buehler <melanie.h.buehler@intel.com>

mhbuehler committed:
Merge pull request #15 from mhbuehler/melanie/port_placement

Fix port number placement

okhleif-10 reviewed (commented):
LGTM, do all tests pass?

mhbuehler commented on comps/cores/mega/gateway.py:
```diff
@@ -988,8 +1036,9 @@ async def handle_request(self, request: Request):
-        # Multimodal RAG QnA With Videos has not yet accepts image as input during QnA.
         num_messages = len(data["messages"]) if isinstance(data["messages"], list) else 1
+
+        # Multimodal RAG QnA With Videos has not yet accepts image as input during QnA.
```
Can you remove this comment or update it for accuracy?

mhbuehler commented on tests/cores/mega/test_multimodalqna_gateway.py:
```diff
@@ -205,9 +222,70 @@ async def test_handle_request(self):
+                self.assertEqual(len(b64_types["image"]), 2)
+            finally:
+                test_gateway.stop()
+
```
Nice test, it passed for me

dmsuehir committed:
Remove old comment and added more accurate description

Signed-off-by: dmsuehir <dina.s.jones@intel.com>

dmsuehir committed:
add comment in code about MAX_IMAGES

Signed-off-by: dmsuehir <dina.s.jones@intel.com>

dmsuehir committed:
Add Gaudi support for image query

Signed-off-by: dmsuehir <dina.s.jones@intel.com>

dmsuehir committed:
Merge branch 'mmqna-audio-query' of github.com:mhbuehler/GenAIComps into dina/image_query

dmsuehir committed:
Merge branch 'mmqna-image-query' of github.com:mhbuehler/GenAIComps into dina/image_query

dmsuehir committed:
Fix to pass the retrieved image last

Signed-off-by: dmsuehir <dina.s.jones@intel.com>

dmsuehir committed:
Revert out gateway and gateway test code, due to its move to GenAIExamples

Signed-off-by: dmsuehir <dina.s.jones@intel.com>

dmsuehir commented:
I rebased and synced this branch with `mmqna-image-query` since the audio input PR has been merged. I also moved the gateway updates to GenAIExamples, and reran the microservice tests for LVM, embddings multimodal, and retriever multimodal redis.

mhbuehler commented on tests/retrievers/test_retrievers_multimodal_redis_langchain.sh:
```diff
@@ -58,6 +58,32 @@ function validate_microservice() {
+
+        if echo "$CONTENT" | grep -q "retrieved_docs"; then
+            echo "[ retriever ] Content has retrieved_docs as expected."
+            if echo "$CONTENT" | grep -q "retrieved_docs"; then
```
Is this grep supposed to look for `img_b64_str`?

dmsuehir replied:
Yes, good point. I will fix this.

dmsuehir committed:
Fix retriever test for checking for b64_img_str in the result

Signed-off-by: dmsuehir <dina.s.jones@intel.com>

mhbuehler reviewed (approved):
LGTM!

okhleif-10 commented on comps/lvms/llava/lvm.py:
```diff
@@ -28,6 +28,9 @@
 logflag = os.getenv("LOGFLAG", False)
 
+# The maximum number of images that should be sent to the LVM
+max_images = int(os.getenv("MAX_IMAGES", 1))
```
In this line, is `1` being set manually? Or is it a default?

dmsuehir replied:
This means that if `MAX_IMAGES` is unset, it will default to 1
